### PR TITLE
Language runners: Better interpreter finding, interpreters configurable, race condition fixed

### PR DESCRIPTION
**Pull request by JanDiederich**
I improved the script language running further.

### Better interpreter finding
I improved the interpreter finding. For example: JavaScript's `node`.<br>
If `node` isn't on the path it can search NVM installations. But... if the default script launchers aren't found, now you can specify the script launcher location yourself, for all the new languages. Or maybe it gets found, but you want to tell it to use another version. That's now possible.

#### Go
I searched again and again, and found that Go seems to be widely used in the setup and management of projects, so I added it.

#### TypeScript
When already using a scripting language, never intended for tasks like file / project management, you want it to at least have types. So I converted the JavaScript scripts of my projects to TypeScript scripts. And with JavaScript and TypeScript I learned the annoying truth: I had to make the execution configurable, because otherwise they won't run as intended.
And: Under Windows TypeScript runs only reliably when using a temp-file. So I added a "run in temp-file" feature too, but made it available to all languages, because that's great for debugging.

### Language interpreter configurable
As said, I made the interpreters configurable.
Go to the plugin settings page, and click on the bottom button.
There you can select the interpreter location, add parameters, and it will even warn you if you use obviously wrong combinations (no annoying message boxes, just a label turning red).

### Rare race-condition fixed when running a script
That was the first thing I fixed, that it was rarely crashing when running a script, but it happened from time to time, with some race condition. So, with that fix, I now have it run for weeks, without any problems.

## What kind of change does this PR introduce?

- [x] Bugfix
- [x] Feature
- [ ] Documentation
- [ ] Other, please describe:

## Does this PR introduce a breaking change?

- [x] No
- [ ] Yes
## Checklist:

- [x] I have updated the [documentation](https://github.com/asciidoctor/asciidoctor-intellij-plugin/blob/main/doc/users-guide/modules/ROOT/pages/features/) to reflect the changes introduced by this PR.

Please, tell me what and if you need some changes.

### CODE REVIEW

**Comment by ahus1** [comment]
@JanDiederich - thank you for this pull request, and sorry for the wait! I fixed a compile problem, and rebased due to a conflict. 

Can you please review the copilot findings above?

**Comment by ahus1** [comment]
@JanDiederich - I see this PR adds Lombok. As there are only very few uses in this PR, can you please implement it without Lombok, of elaborate why it should be used?

File: src/main/java/org/asciidoc/intellij/commandRunner/AsciiDocRunner.java [hunk 1]
```diff
@@ -6,12 +6,11 @@
 import com.intellij.openapi.project.Project;
 import com.intellij.openapi.vfs.VirtualFile;
 import org.jetbrains.annotations.Nls;
-import org.jetbrains.annotations.NotNull;
 
 public interface AsciiDocRunner {
   ExtensionPointName<AsciiDocRunner> EP_NAME = ExtensionPointName.create("org.asciidoc.intellij.asciidocRunner");
 
-  boolean isApplicable(@NotNull Project project, Language language);
+  boolean isApplicable(Language language);
```
[the remark this answers is not in the document]
Comment (JanDiederich) [reply src/main/java/org/asciidoc/intellij/commandRunner/AsciiDocRunner.java:13]: @ahus1 That's seems like a typical AI comment. I was the one who introduced that `Project` parameter, w/o any problems. Now I can remove it again. It became superfluous.

File: src/main/java/org/asciidoc/intellij/commandRunner/arbitrary/AsciiDocRunnerArbitrary.java [hunk 2]
```diff
@@ -116,30 +194,52 @@ private void runCommand(Project project, GeneralCommandLine commandLine, Console
...
    *
-   * @param project Project.
    * @return Interpreter.
    */
-  abstract String findInterpreter(Project project);
+  @Nullable
+  abstract String findInterpreter();
+
+  @Nullable
+  abstract AsciiDocScriptLanguageSetting extractScriptLanguageSetting(AsciiDocScriptLanguageSettings languageSettings);
```
[the remark this answers is not in the document]
Comment (JanDiederich) [reply src/main/java/org/asciidoc/intellij/commandRunner/arbitrary/AsciiDocRunnerArbitrary.java:222]: @ahus1 Again, that seems like a non-problem to me, that is basically an internal class. That the bot complains about adding non-existing new hooks is another problem.
In general it seems it doesn't like changing interfaces or abstract methods, as if that were all external API changes, that seems absurd.

File: src/main/java/org/asciidoc/intellij/settings/AsciiDocPreviewSettings.java [hunk 3]
```diff
@@ -125,7 +132,8 @@ public AsciiDocPreviewSettings(@NotNull SplitFileEditor.SplitEditorLayout splitE
                                  boolean enableBuiltInMermaid,
                                  int zoom,
                                  boolean hideErrorsInSourceBlocks,
-                                 @Nullable String hideErrorsByLanguage) {
+                                 @Nullable String hideErrorsByLanguage,
+                                 @Nullable AsciiDocScriptLanguageSettings scriptLanguageSettings) {
```
[the remark this answers is not in the document]
Comment (JanDiederich) [reply src/main/java/org/asciidoc/intellij/settings/AsciiDocPreviewSettings.java:136]: @ahus1 I think that was the compile error you fixed? Can't see an error there (anymore?).

File: src/main/java/org/asciidoc/intellij/settings/AsciiDocPreviewSettingsForm.java [hunk 4]
```diff
@@ -94,6 +100,24 @@ public class AsciiDocPreviewSettingsForm implements AsciiDocPreviewSettings.Hold
...
+  public AsciiDocPreviewSettingsForm() {
+    setupLanguages.addActionListener(actionEvent -> showScriptLanguagesForm());
+  }
+
+  private void showScriptLanguagesForm() {
+    AsciiDocScriptLanguagesForm asciiDocScriptLanguagesForm = new AsciiDocScriptLanguagesForm(myMainPanel,
+      myScriptLanguageSettings,
+      (@NotNull AsciiDocScriptLanguageSettings acceptedSettings) -> {
+        myScriptLanguageSettings = acceptedSettings;
+        log.info("Set language interpreter settings to %s.".formatted(myScriptLanguageSettings));
```
[the remark this answers is not in the document]
Comment (JanDiederich) [reply src/main/java/org/asciidoc/intellij/settings/AsciiDocPreviewSettingsForm.java:117]: @ahus1 But it doesn't contain secretive values, and is very helpful during bug hunting. What I suggest is, setting the log call from `info` to `debug`.

File: src/main/java/org/asciidoc/intellij/commandRunner/arbitrary/AsciiDocSuggestedParameter.java [hunk 5]
```diff
@@ -0,0 +1,18 @@
+package org.asciidoc.intellij.commandRunner.arbitrary;
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.io.Serializable;
+
+/**
+ * Suggested parameters that can be very helpful to pass, everytime the interpreter runs.
```
[the remark this answers is not in the document]
Comment (JanDiederich) [reply src/main/java/org/asciidoc/intellij/commandRunner/arbitrary/AsciiDocSuggestedParameter.java:9]: @ahus1 I actually prefer the shortened version.

**Comment by JanDiederich** [comment]
@ahus1 Although I use Lombok in "only" 9 classes, it saves me big amounts of boilerplate code.

Take `AsciiDocScriptLanguageSettings` for example, where I use `@NoArgsConstructor` (saves me only 3 lines), but then `@AllArgsConstructor`, which saves me a chatty constructor w/ 7 params, Then there is all the code from `@Data` and so on. **Removing** the Lombok annotations **would blow up the class from small 65 lines to whopping 254**, with only the tiny benefit of one small removed dependency.

There there's `AsciiDocLanguageSettingsParam`, if I remove the `@Builder` annotation the class would go from 28 lines up to 73.

Other classes aren't that dramatic, but it still adds up.

And: I've seen Lombok in big, professional corporate environments used again and again, so even in that restrict environments, I can't see a problem.

File: src/main/java/org/asciidoc/intellij/commandRunner/arbitrary/SplitNodeJsVersion.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+      int numA = i < numbers.length() ? numbers.get(i) : 0;
+      int numB = i < other.numbers.length() ? other.numbers.get(i) : 0;
+      if (numA != numB) {
+        // Descending.
+        return numB - numA;
+      }
+    }
+    if (suffix != null) {
+      if (other.suffix != null) {
+        return suffix.compareToIgnoreCase(other.suffix) * -1;
```
[the remark this answers is not in the document]
Comment (JanDiederich) [reply src/main/java/org/asciidoc/intellij/commandRunner/arbitrary/SplitNodeJsVersion.java:53]: @ahus1 While true, in the last 17 years of Node.JS there have been only [10 times even a "rc-4"](https://nodejs.org/download/rc/index.json), which is the highest "rc" number ever in Node.JS. I advice against writing that code, I don't think it's worth it. Especially since it's surprisingly complex, and will most likely never be used. And even if that failed, the user can always set another version him-/herself.

File: src/main/java/org/asciidoc/intellij/commandRunner/arbitrary/AsciiDocRunnerForGo.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+  }
+
+  public static boolean isGo(String value) {
+    return value.equalsIgnoreCase("go");
+  }
+
+  @NotNull
+  public static String findGoInterpreter() {
+    return SystemInfo.isWindows ? WINDOWS_EXECUTABLE : UNIX_EXECUTABLE;
+  }
```
[the remark this answers is not in the document]
Comment (JanDiederich) [reply src/main/java/org/asciidoc/intellij/commandRunner/arbitrary/AsciiDocRunnerForGo.java]: @ahus1 Actually a valid point. I changed that.

File: src/main/java/org/asciidoc/intellij/commandRunner/arbitrary/SplitNodeJsVersion.java [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+      } else {
+        return 1;
+      }
+    } else if (other.suffix != null) {
+      return 1;
```
[the remark this answers is not in the document]
Comment (JanDiederich) [reply src/main/java/org/asciidoc/intellij/commandRunner/arbitrary/SplitNodeJsVersion.java]: @ahus1 The order in this edge-case isn't that important, I just followed the AI here and changed the expected value in the unit test accordingly.

**Comment by JanDiederich** [comment]
I rebased on main and force-pushed that.

File: src/main/java/org/asciidoc/intellij/commandRunner/arbitrary/AsciiDocRunnerForTypeScript.java [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+  @Nullable List<String> codeRunParameters(@Nullable AsciiDocScriptLanguageSetting languageSetting) {
+    final List<String> strings;
+    if (SystemInfo.isWindows) {
+      String utilPath = Optional.ofNullable(
+          AsciiDocApplicationSettings.getInstance().getAsciiDocPreviewSettings().getScriptLanguageSettings())
+        .map(AsciiDocScriptLanguageSettings::getLanguageSettingTypeScript)
+        .map(AsciiDocScriptLanguageSetting::getUtilPath)
+        .orElse(null);
+      String npxScript;
+      if (utilPath == null) {
```
[the remark this answers is not in the document]
Comment (JanDiederich) [reply src/main/java/org/asciidoc/intellij/commandRunner/arbitrary/AsciiDocRunnerForTypeScript.java]: @ahus1 Valid point, I fixed that.

File: src/main/java/org/asciidoc/intellij/commandRunner/arbitrary/AsciiDocRunnerArbitrary.java [hunk 10]
```diff
@@ -50,34 +74,83 @@ record ConsoleData(@NotNull ConsoleView consoleView,
...
       return false;
     }
+    // Add the code, what this is actually all about, at last.
+    final File tempFile;
+    if (useTempFile) {
+      // Use no inline evaluation, run a code-file.
+      TempFileInfo tempFileInfo = getTempFileInfo();
+      try {
+        tempFile = File.createTempFile(virtualFile.getName(), tempFileInfo.extension());
+        tempFile.deleteOnExit();
```
[the remark this answers is not in the document]
Comment (JanDiederich) [reply src/main/java/org/asciidoc/intellij/commandRunner/arbitrary/AsciiDocRunnerArbitrary.java:107]: @ahus1  You can't run hidden files w/ go, meaning no `.ad` files can be used as source (and even that would be 3 characters long).
So even if the source file is named `x.ad`, the shortest AsciiDoc extension available, it will be > 3 characters long.
But if you want to be sure, I can add that defensive code?

File: src/main/java/org/asciidoc/intellij/commandRunner/arbitrary/AsciiDocBackgroundCommand.java [hunk 11]
```diff
@@ -28,8 +28,8 @@ class AsciiDocBackgroundCommand extends Task.Backgroundable {
   private final AsciiDocRunnerArbitrary.ConsoleData consoleData;
 
   @Nullable
-  private ProcessHandler processHandler;
-  private boolean abort = false;
+  private volatile ProcessHandler processHandler;
+  private volatile boolean abort = false;
```
[the remark this answers is not in the document]
Comment (JanDiederich) [reply src/main/java/org/asciidoc/intellij/commandRunner/arbitrary/AsciiDocBackgroundCommand.java:39]: The `abort` flag here isn't a signal for a thread to stop its execution loop. It's there to inform the thread loop that it was intentionally aborted by a user action. Not by a normal finished script, or an error. So the thread knows to now show an "Unexpected exit" message or something like that.

I added a JavaDoc explaining this directly on the variable to stop misunderstandings, because the "Loop-Stopping abort flag" pattern is a known pattern (a pattern which isn't used here, instead a specific IntelliJ mechanism is used).

**Comment by JanDiederich** [comment]
@ahus1 I worked through all AI comments. Awaiting your opinions.
